cpp/oneapi/dal/backend/primitives/selection/kselect_by_rows_heap_dpc.cpp

```diff
@@ -0,0 +1,448 @@
+
+public:
+    kernel_select_heap() = default;
+
+    template <bool ids_only = (ids_out && !dst_out), typename = std::enable_if_t<ids_only>>
+    kernel_select_heap(const dst_t* const src_ptr,
+                       idx_t* const ids_ptr,
+                       std::int32_t k,
+                       std::int32_t width,
+                       std::int32_t height,
+                       std::int32_t src_str,
+                       std::int32_t ids_str,
+                       acc_t heaps)
+            : base_t(ids_str, ids_ptr),
+              src_str_(src_str),
+              src_ptr_(src_ptr),
+              k_(k),
+              width_(width),
+              height_(height),
+              heaps_(heaps) {}
+
+    template <bool dst_only = (dst_out && !ids_out), typename = std::enable_if_t<dst_only>>
+    kernel_select_heap(const dst_t* const src_ptr,
+                       dst_t* const dst_ptr,
+                       std::int32_t k,
+                       std::int32_t width,
+                       std::int32_t height,
+                       std::int32_t src_str,
+                       std::int32_t dst_str,
+                       acc_t heaps)
+            : base_t(dst_str, dst_ptr),
+              src_str_(src_str),
+              src_ptr_(src_ptr),
+              k_(k),
+              width_(width),
+              height_(height),
+              heaps_(heaps) {}
+
+    template <bool both = (ids_out && dst_out), typename = std::enable_if_t<both>>
+    kernel_select_heap(const dst_t* const src_ptr,
+                       idx_t* const ids_ptr,
+                       dst_t* const dst_ptr,
+                       std::int32_t k,
+                       std::int32_t width,
+                       std::int32_t height,
+                       std::int32_t src_str,
+                       std::int32_t ids_str,
+                       std::int32_t dst_str,
+                       acc_t heaps)
+            : base_t(ids_str, dst_str, ids_ptr, dst_ptr),
+              src_str_(src_str),
+              src_ptr_(src_ptr),
+              k_(k),
+              width_(width),
+              height_(height),
+              heaps_(heaps) {}
+
+    void operator()(sycl::nd_item<1> item) const {
+        using sycl::ONEAPI::reduce;
+
+        auto sg = item.get_sub_group();
+
+        const std::int32_t sid = sg.get_group_linear_id();
+        const std::int32_t cid = sg.get_local_linear_id();
+        const std::int32_t sg_width = sg.get_local_range().size();
+        const std::int32_t rid = sid + item.get_group_linear_id() * sg.get_group_range().size();
+
+        if ((rid >= height_) || (cid >= sg_width))
+            return;
+
+        dst_t pbuff_dst[pbuff_size] = { dst_default };
+        idx_t pbuff_ids[pbuff_size] = { idx_default };
+        std::int32_t pbuff_count, prev_count;
+
+        sel_t* const heaps = heaps_.get_pointer();
+        sel_t* const curr_heap = heaps + (k_ + 1) * sid;
```

Comment:
Number of subgroups should be derived from SLM size and "k"?

Reply:
It's actually kernel code. But yes it can be derived and it is done in wrapper